SQL: Enable adding missing equals to bool fields as filters

bpintea opened the pull request:
This PR adds QL's `AnalyzerRules.AddMissingEqualsToBoolField` to SQL's analyser as well.

Closes #65579.

bpintea committed:
Enable the AddMissingEqualsToBoolField rule in SQL

This will enable QL's AnalyzerRules.AddMissingEqualsToBoolField to SQL's
analyzer as well.

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/analyzer/AnalyzerRules.java:
```diff
@@ -39,7 +40,7 @@ protected LogicalPlan rule(Filter filter) {
 
         private Expression replaceRawBoolFieldWithEquals(Expression e) {
-            if (e instanceof FieldAttribute) {
+            if (e instanceof FieldAttribute && e.dataType() == BOOLEAN) {
```
The rule could check whether the `Filter` is resolved before applying the equality.

bpintea replied:
Ah, right. Thanks, fixed.

costin reviewed (approved):
LGTM

bpintea committed:
Fix: check type resolution before type's val

Make sure a fields's type is available before checking it's value.

palesz reviewed (commented):
LGTM.

So the outcome for SQL: Before all these queries below failed with exception during the query translation, while after this PR the first one will succeed.

```sql
SELECT * FROM test WHERE bool
SELECT * FROM test WHERE int
SELECT * FROM test WHERE str
```

bpintea commented:
> So the outcome for SQL: Before all these queries below failed with exception during the query translation, while after this PR the first one will succeed.

That's correct.

bpintea committed:
Rework the way resolution is checked

Check on filter's resolution part of the rule() method, before any other
change is applied.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/add_missing_eq_to_bool

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:
```diff
@@ -2459,4 +2461,33 @@ public void testReplaceSumWithStats() {
     }
+
+    public void testAddMissingEqualsToBoolField() {
+        LogicalPlan p = plan("SELECT bool FROM test WHERE bool");
```
Maybe I'm missing some context, but what happens to nested expressions like: `WHERE a > 1 and (bool or b < 2) or (c = 3 and bool)`, could we have a test for that as well?

bpintea replied:
They'd be eval'd to a `BOOLEAN` type, so nothing would happen: the `AddMissingEqualsToBoolField()` wouldn't change the expression (since its root node type wouldn't be a FieldAttribute) and the verifier check would let it pass (since it's a bool type). Basically most of the tests we have look to these two newly added checks similar to your example, but I could still add one if you think it'd help.

bpintea replied:
@matriv, I've added a slightly extended version of your test (and updated my first eval of the case above).

matriv reviewed (commented):
Left a comment.

bpintea committed:
Add test with nested bool

Add one SQL test with nested "free-floating" bool that got assigned a
true literal.

bpintea committed:
Merge branch 'enh/add_missing_eq_to_bool' of github.com:bpintea/elasticsearch into enh/add_missing_eq_to_bool

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enh/add_missing_eq_to_bool

bpintea committed:
Style fix

- minor style fix.

matriv reviewed (approved):
LGTM, thx!
